cpp/src/arrow/result.h

```diff
@@ -371,7 +371,7 @@ class [[nodiscard]] Result : public util::EqualityComparable<Result<T>> {
     if (ok()) {
```

Comment:
Additionally, I don't think this should be `G&&` as the intention here is to let the compiler not even materialize `G` (e.g. by allocating an object on the stack and calling some constructor for `G`), but instead just splice the code from the lambda into the inlined code for this function.

This is how predicates are passed to functions like `find_if` in the standard library.

```cpp
template< class InputIt, class UnaryPredicate >
constexpr InputIt find_if( InputIt first, InputIt last, UnaryPredicate p );
```

Reply:
> Additionally, I don't think this should be `G&&` as the intention here is to let the compiler not even materialize `G` (e.g. by allocating an object on the stack and calling some constructor for `G`), but instead just splice the code from the lambda into the inlined code for this function.

I'm not sure what you mean here. `G&&` is a [forwarding reference](https://en.cppreference.com/w/cpp/language/reference#Forwarding_references) according to the C++ reference.

> This is how predicates are passed to functions like `find_if` in the standard library.

Is this the actual declaration in a standard library implementation, or the abstract signature as shown in the reference?
(also, since `find_if` typically calls the predicate several times, it can probably not `std::forward` it)

Reply:
> > Additionally, I don't think this should be `G&&` as the intention here is to let the compiler not even materialize `G` (e.g. by allocating an object on the stack and calling some constructor for `G`), but instead just splice the code from the lambda into the inlined code for this function.
> 
> I'm not sure what you mean here. `G&&` is a [forwarding reference](https://en.cppreference.com/w/cpp/language/reference#Forwarding_references) according to the C++ reference.

Yes, but why do we want to forward `G` when we want the compiler to completely inline the code it represents?

> 
> > This is how predicates are passed to functions like `find_if` in the standard library.
> 
> Is this the actual declaration in a standard library implementation, or the abstract signature as shown in the reference? 

Yes. This is a common way of passing lambdas that you want to be completely inlined if they are inline at the callsite.

See https://github.com/llvm/llvm-project/blob/main/libcxx/include/__algorithm/find_if.h#L23

> (also, since `find_if` typically calls the predicate several times, it can probably not `std::forward` it)

Again, I don't understand why we should forward it. We only call it, don't store/pass it anywhere.

Another example from the STL is how `std::sort` inlines `Compare __c` everywhere. If you pass a lambda it will inline the lambda, if you pass a function pointer or an `std::function` it will invoke it by following the pointer and so on.

Reply:
This is becoming pedantic. Do we want to solve #33911 or not?

Reply:
(and allow for https://github.com/apache/arrow/pull/33912#issuecomment-1409091955)

Reply:
Sorry. In my other comment I gave a suggestion on how to solve the issue.

Proof that it works as intended: https://godbolt.org/z/5c17zjG89

Reply:
My full proposal is (1) Remove the `&&` from `G` and (2) add the `enable_if` requirement.

Reply:
> Proof that it works as intended: https://godbolt.org/z/5c17zjG89

I'm a little confused.  In this example I get:

```
function &:
copy
copy

function && (with std::move):
move
move
```

If I change to `G&&` I get:

```

function &:

function && (with std::move):
```

It seems the version without the moves (and especially without the copies) would be preferred.

Also, if your generator is stateful (e.g. a generator object that returns the next id in a sequence) then it would make a big difference if you use & or &&.

I agree, in this case, it seems rather unlikely that someone is going to be passing in a stateful (or complex to copy, can't be inlined) generator.   Still, I don't think there is any penalty to the &&.  And the standard [does sometimes pass functions by forwarding reference](https://en.cppreference.com/w/cpp/utility/functional/invoke).

Reply:
[An even more applicable example](https://en.cppreference.com/w/cpp/utility/optional/or_else)

Reply:
> It seems the version without the moves (and especially without the copies) would be preferred.

You're right @westonpace. I stand corrected.

The implementation from `optional` you linked does exactly as this PR plus a `static_assert` with a nice error message that we could adopt.

Reply:
Thank for your reviews and suggestions!

But I want to point out that there are some difference between `Result::ValueOrElse` and `std::optional::or_else`: `Result<T>::ValueOrElse` returns a value typed `T`, while `std::optional<T>::or_else` typed `std::optional<T>`, so that the generator in `std::optional::or_else` can returns something like `std::nullopt` and then the `or_else` operations can be "chained" (e.g. `v.and_then().or_else().transform()`) and more expressive to achieve "monadic".